### ISSUE TITLE
Potential memory leak issue caused by new changes to *_drbg_free

### ISSUE THREAD

**Issue by zhenghaven** [ARMmbed/mbedtls#4045]
### Description
- Type: Bug

---------------------------------------------------------------
## Bug

**mbed TLS build:**  
Version: 2.25.0 or git commit 6aceb51 

Hi,

I'm trying to upgrade from v2.24.0 to v2.25.0 and found a potential memory leak issue during my unit testing.

The new changes to `mbedtls_*_drbg_free()` function allows the drbg context to be reset to the state to just after `mbedtls_*_drbg_init()`. But if I enable multi-threading with `MBEDTLS_THREADING_C`, `mbedtls_*_drbg_free()` function will call `mbedtls_mutex_free()` to free the mutex and call `mbedtls_mutex_init()` to init it again. So `mbedtls_*_drbg_free()` will never actually free the mutex, and if there is memory allocation happens inside `mbedtls_mutex_init()`, this will cause the memory leak on the mutex.

**Comment by Christian-Sander** [comment]
I can confirm this issue, it just broke my mutex implementation because of a mutex stored on stack in ecp_mul_comb() that isn't freed properly, leaving invalid data in my mutex linked list. For now I will change my mbedTLS threading abstraction to lazy initialization the first time a mutex is locked.

**Comment by gilles-peskine-arm** [comment]
There's a proposed fix in https://github.com/ARMmbed/mbedtls/pull/4090. Please note that it has not been reviewed yet, and we may yet revise how we resolve this issue (it's not fully clear to me how to fix this issue without breaking “unusual” usage of the API). If you want to try it out, you can apply the patches from https://github.com/gilles-peskine-arm/mbedtls/tree/fix-drbg-mutex-usage-2.25-untested.

**Comment by gilles-peskine-arm** [comment]
I'm closing this issue as a duplicate of https://github.com/ARMmbed/mbedtls/issues/4017. It's either exactly the same problem (maybe in a different branch) or a slight variation (e.g. HMAC_DRBG vs CTR_DRBG) which will be fixed together. Please reopen this issue once the proposed fix is merged if it doesn't fix your problem.

**Comment by gilles-peskine-arm** [comment]
The fix is now merged into the supported branches (https://github.com/ARMmbed/mbedtls/tree/mbedtls-2.7, https://github.com/ARMmbed/mbedtls/tree/mbedtls-2.16 and https://github.com/ARMmbed/mbedtls/tree/development). The critical parts (mutex resource leak through a DRBG context) is identical to the patches I shared earlier.
